src/imports/background/import-item-creation.js

```diff
@@ -0,0 +1,195 @@
+// Binds old ext bookmark urls to a function that transforms old ext data to an import item
+const transformOldExtDataToImportItem = bookmarkUrls => (item, index) => ({
+    type: IMPORT_TYPE.OLD,
+    url: item.url,
+    timestamp: item.time,
+    hasBookmark: bookmarkUrls.has(item.url),
+    index,
+})
+
+// Binds an import type to a function that transforms a history/bookmark doc to an import item.
+const transformBrowserItemToImportItem = type => item => ({
+    browserId: item.id,
+    url: item.url,
+    // HistoryItems contain lastVisitTime while BookmarkTreeNodes contain dateAdded
+    timestamp: item.lastVisitTime || item.dateAdded,
+    type,
+})
+
+async function checkWithExistingBookmarks() {
+    const { rows: existingBmDocs } = await db.allDocs({
+        startkey: bookmarkKeyPrefix,
+        endkey: `${bookmarkKeyPrefix}\uffff`,
+    })
+
+    const existingIds = new Set(
+        existingBmDocs.map(row => convertBookmarkDocId(row.id).url),
+    )
+    return url => existingIds.has(url)
+}
+
+/**
+ * @returns {({ url: string }) => boolean} A function affording checking of a URL against the
+ *  URLs of existing page docs.
+ */
+async function checkWithExistingPages() {
+    const { rows: existingPageDocs } = await db.allDocs({
+        startkey: pageKeyPrefix,
+        endkey: `${pageKeyPrefix}\uffff`,
+    })
+
+    const existingIds = new Set(
+        existingPageDocs.map(row => convertPageDocId(row.id).url),
+    )
+    return url => existingIds.has(url)
+}
+
+async function initFilterItemsByUrl() {
+    const isNotBlacklisted = await checkWithBlacklist()
+
+    /**
+     * Performs all needed filtering on a collection of history, bookmark, or old ext items. As these
+     * can be quite large, attemps to do in a single iteration of the input collection.
+     *
+     * @param {Array<any>} items Array of items with `url` to filter on.
+     * @param {(item: any) => any} transform Opt. transformformation fn turning current iterm into import item structure.
+     * @param {(url: string) => bool} alreadyExists Opt. checker function to check against existing data.
+     * @return {Map<string, any>} Filtered version of `items` array made into a Map of encoded URL strings to import items.
+     */
+    return (items, transform = f => f, alreadyExists) =>
+        items.reduce((accItems, currItem) => {
+            if (!isLoggable(currItem) || !isNotBlacklisted(currItem)) {
+                return accItems
+            }
+
+            try {
+                // Augment the item with encoded URL for DB check and item map
+                const encodedUrl = encodeUrl(currItem.url, false)
+
+                // Filter out items already existing in DB, if specified
+                if (alreadyExists == null || !alreadyExists(encodedUrl)) {
+                    accItems.set(encodedUrl, transform(currItem))
+                }
+                return accItems
+            } catch (error) {
+                // URI malformed; don't add to list
```

Comment:
is this the catch only for encodeUrl or the transform can fail too?

Reply:
I generally don't care much about sub-optimization but since this is a refactor for speed it's worth trying to reduce the try-catch size (https://github.com/petkaantonov/bluebird/wiki/Optimization-killers)

Reply:
This article looks interesting 👍 only the `encodeUrl` can fail. `transform` is just a pure function mapping source data to the imports data shape, or identity function by default.

Reply:
as above